### PR TITLE
allocator: add support for store pool liveness overrides

### PR DESCRIPTION
**Pull request by AlexTalks**
While previously the allocator only evaluated using liveness obtained
from gossip, this change introduces a new `OverrideStorePool` struct
which can be used to override the liveness of a node for the purposes of
evaluating allocator actions and targets.  This `OverrideStorePool` is
backed by an existing actual `StorePool`, which retains the majority of
its logic.

Depends on #91461.

Part of #91570.

Release note: None

### CODE REVIEW

**Review by kvoli — APPROVED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91965)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks and @andrewbaptist)

___
*[`pkg/kv/kvserver/allocator/storepool/override_store_pool.go` line 29 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91965#-NI43Fmm5WxbASgAPgnn:-NI43Fmm5WxbASgAPgno:bjcm33j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e79788d4f5ab41222011307299cf05b0189479f6/pkg/kv/kvserver/allocator/storepool/override_store_pool.go#L29)):*
> ```Go
> // The OverrideStorePool's methods fall through to the underlying StorePool
> // while passing in the configured NodeLivenessFunc to be used in place of the
> // underlying StorePool's. This is only the case, however, for StorePool methods
> ```

nit: This last sentence is hard to parse - could it be reworded to be clearer?



_Code quote:_
```Go
// underlying StorePool's. This is only the case, however, for StorePool methods
// that are read-only, despite StorePool.DetailsMu being held in write mode in
// some functions, avoiding the possibility of mutating underlying state.
```
___
*[`pkg/kv/kvserver/allocator/storepool/override_store_pool_test.go` line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91965#-NI4ASLt8wIf_7W3IDl5:-NI4ASLt8wIf_7W3IDl6:bieqeuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e79788d4f5ab41222011307299cf05b0189479f6/pkg/kv/kvserver/allocator/storepool/override_store_pool_test.go#L179)):*
> ```Go
> 	liveReplicas, deadReplicas := sp.LiveAndDeadReplicas(replicas, false /* includeSuspectAndDrainingStores */)
> 	if len(liveReplicas) != 5 {
> 		t.Fatalf("expected five live replicas, found %d (%v)", len(liveReplicas), liveReplicas)
> ```

Why not use the `require` lib in these tests?

**Review by AlexTalks — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91965)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @kvoli)

___
*[`pkg/kv/kvserver/allocator/storepool/override_store_pool_test.go` line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91965#-NI4ASLt8wIf_7W3IDl5:-NItCs6y3X6Jo4laBDkD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e79788d4f5ab41222011307299cf05b0189479f6/pkg/kv/kvserver/allocator/storepool/override_store_pool_test.go#L179)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Why not use the `require` lib in these tests?
</blockquote></details>

Done.
